src/restLink.ts

```diff
@@ -393,15 +407,26 @@ const resolver: Resolver = async (
     }
 
     validateRequestMethodForOperationType(method, operationType || 'query');
-
-    return await (customFetch || fetch)(`${uri}${pathWithParams}`, {
+    const requestParameters = {
       credentials,
       method,
       headers,
       body,
-    })
-      .then(res => res.json())
-      .then(result => addTypeNameToResult(result, type));
+    };
+    let requestResult: Promise<any> = null;
+
+    if (customRequest) {
+      requestResult = customRequest(requestUrl, requestParameters);
+    } else {
+      requestResult = (customFetch || fetch)(
+        requestUrl,
+        requestParameters,
+      ).then(res => res.json());
```

Comment:
Looking at all your code changes, they all appear to be unnecessary. Instead you could wrap your custom requestor so that it becomes more fetch-compatible.

```
// Assuming you have: myGetJSON() => Promise<{}>

/** Wraps myGetJSON so that it becomes customFetch-compatible! */
const MyGetJSON_FetchCompatible = (url:string | Request, requestParameters:…) => Promise<…> {
    return myGetJSON(url, requestParameters).then(res => {
        return {
            json: () => res
        }
    });
}
```

Reply:
The thing is, I don't really think it's a good idea to use Response as a response interface.
Maybe it's even better to separate data-fetching fully. Otherwise you are creating a weird contract, you're expecting to get a JSON, which then will be mutated with according type names, but in fact you are waiting for Response object.
My point - decouple network interaction and data mutation. So, that you can provide a reasonable interface for your network layer, which then can be implemented in all possible ways.

Reply:
I hear you @VladimirMilenko. I do worry about the coupling, but I think we probably need a solution here that is more holistic with the rest of ApolloGraphQL.

I'd support the `customFetch` parameter type changing to `<FetchType> | <NewGenericFetchInterface>` but I think that `NewGenericFetchInterface` probably should be defined at the ApolloGraphQL/ApolloClient-Core level.

My fear is that the current suggestion is a lot of extra surface area, and puts two different, *almost* identical options on our configuration. That feels like heavy cost for otherwise simple behavior forking. -- Maybe other people can chime in on their preferences?

Reply:
I was not talking about going same way, as in this PR. I was talking about:
1) Make a network interface
2) Declare the basic one, which will use a fetch, or customFetch.
3) Let others create own adapters:
```
const link = new RestLink({
networkAdapter: myVeryOwnAdapter
});
```

Reply:
Think of it as a convenient solution.
Someone wants to use axios instead of fetch - add AxiosNetworkAdapter.
And that's it. I might create a PR for this, but I want to hear your opinion

Reply:
OH! -- Then I totally agree with you. I think the networkAdapter generic interface would be great. My point though is that we should probably coordinate this with apolloclient / core team. I think having the definition of that generic network adapter inside `apollo-link-rest` might be the wrong place. Does that make sense?

Reply:
Don't think so, it's very package-dependent.
You can't apply same adapter to all apollo-links, cuz, apollo-link is already like a network adapter.
But, for rest - that's actually a good thing. Cuz you can even outsource query building, requests making.
It can be more solid, cuz you can allow to add dynamic parameters to your query.
Like: ```@rest(contentType: 'form')```.
And even path processing and etc.
This all can be outsourced to developers.

Reply:
@VladimirMilenko can you detail a bit more `networkAdapter` parameter ?
I think the network interface is also a good idea, but I do not get why this package should define it. This is something much more generic.

Reply:
@VladimirMilenko Maybe your could adapt your PR with networkAdapter ?
I tend to agree with frederic that this is not an apollo-link-rest specific problem.
I like this conversation so let's see where we will go with it !